### PR TITLE
dialect/sql/schema: file based type store

### LINKED ISSUE #2587: Versioned migrations do not work with GlobalUniqueID
Suppose trying to generate versioned migrations files for a global unique id enabled schema: 

```go
client.Schema.Diff(ctx, schema.WithDir(dir), schema.WithGlobalUniqueID(true))
```

will result into the following error:

```text
insert into ent_types: Error 1146: Table 'migrate.ent_types' doesn't exist
```

- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/ent/ent/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

There is an error.

## Expected Behavior 🤔

A new set of migration files are created that also allocate the correct ranges for the primary keys.

### PR DESCRIPTION
**Pull request by masseelch**
This PR adds support for a file based type storage when using versioned migrations. The file called `.ent_types` is written to the migration directory alongside the migration files and will be kept in sync for every migration file generation run.

In order to not break existing code, where the type storage might differ for different deployment, global unique ID mut be enabled by using a new option. This will also be raised as an error to the user when attempting to use versioned migrations and global unique ID.

Documentation will be added to this PR once feedback on the code is gathered.

Fixes #2587 

----- 

I have discussed with Ariel already, that we will move the configuration for allocated pk ranges to the schema level. We haven't yet figured out, how to do this best, but it will come.

### CODE REVIEW

File: dialect/sql/schema/atlas.go [hunk 1]
```diff
@@ -326,6 +347,18 @@ func (m *Migrate) setupAtlas() error {
...
 	}
+	if m.universalID && m.atlas.dir != nil {
+		// If global unique ids and a migration directory is given, enable the file based type store for pk ranges.
+		m.typeStore = &dirTypeStore{dir: m.atlas.dir}
+		// The file based pk range store is not backwards compatible, since the allocated ranges were computed
+		// dynamically when computing the diff between a deployed database and the current schema. In cases where there
+		// exist multiple deployments, the allocated ranges for the same type might be different from each other,
+		// depending on when the deployment took part. To guard the user against bugs, the file based type store must
+		// be enabled by an option.
+		if !m.atlas.typeStoreConsent {
```
Comment (a8m) [inline dialect/sql/schema/atlas.go:358]: This is a copy of the comment above.

File: dialect/sql/schema/migrate.go [hunk 2]
```diff
@@ -182,11 +183,45 @@ func (m *Migrate) NamedDiff(ctx context.Context, name string, tables ...*Table)
...
 	plan, err := m.atDiff(ctx, m, name, tables...)
 	if err != nil {
 		return err
 	}
+	if m.universalID {
+		// Add the required entries to the TypeTable.
+		for _, t := range m.typeStore.(*dirTypeStore).newTypes {
+			plan.Changes = append(plan.Changes, &migrate.Change{
+				Cmd:     m.atTypeRangeSQL(t),
+				Comment: `add pk range for "t" table`,
```
Comment (a8m) [inline dialect/sql/schema/migrate.go:219]:
```suggestion
				Comment: fmt.Sprintf("add pk range for "%q" table", TypeTable),
```

**Comment by lagzi** [comment]
Hey, I'm getting `validating migration directory: checksum mismatch` when trying to create a new migration (`.Diff(..)`) because it seems that the `.ent_types` file changed but the checksum isn't updated.

**Comment by masseelch** [comment]
> Hey, I'm getting `validating migration directory: checksum mismatch` when trying to create a new migration (`.Diff(..)`) because it seems that the `.ent_types` file changed but the checksum isn't updated.

Good catch. I will both add a test and a fix for that :-)

**Comment by masseelch** [comment]
@lagzi I changed added a commit, that should fix your problem. If it doesn't please ping me on [Discord](https://discord.gg/qZmPgTE6RX) so I can help you resolve it :-)

File: dialect/sql/schema/atlas.go [hunk 3]
```diff
@@ -537,6 +568,46 @@ func (m *Migrate) aIndexes(b atBuilder, t1 *Table, t2 *schema.Table) error {
...
+					return nil, fmt.Errorf("unexpected missing type range for %q", t)
+				}
+			}
+			// Override the types file created in the diff process with the "old" allocated types ranges.
+			if err := m.typeStore.(*dirTypeStore).save(m.dbTypeRanges); err != nil {
+				return nil, err
+			}
+			// Change the type range allocations since they will be added to the migration files when
+			// writing the migration plan to migration files.
+			m.typeRanges = m.dbTypeRanges
```
Comment (a8m) [inline dialect/sql/schema/atlas.go:605]: 💯

File: dialect/sql/schema/atlas.go [hunk 4]
```diff
@@ -574,3 +645,52 @@ func descIndexes(idx *Index) map[string]bool {
...
 	return descs
 }
+
+const entTypes = ".ent_types"
+
+// dirTypeStore stores and read pk information from a text file stored alongside generated versioned migrations.
+// This behaviour is enabled automatically when using versioned migrations.
+type dirTypeStore struct {
+	dir migrate.Dir
+	// newTypes []string // store types added in this run
```
Comment (a8m) [inline dialect/sql/schema/atlas.go:655]: ??

**Comment by a8m** [comment]
Looks good 🚀 

Please, fix the broken tests and I'll approve.

**Comment by masseelch** [comment]
Currently this PR is blocked on a checksum problem: `.ent_types` is added to the checksum as well and this has to be resolved on Atlas.

**Review by a8m — APPROVED** [review]
I like the code but can't see `WithDeterministicGlobalUniqueID`. It's too long for me 🙈

**Pull request merged**